File: app/CerealExercise/CerealExercise/Services/LifetimeStatsCalculator.swift
import Foundation

enum LifetimeStatsCalculator {
    struct Stats: Equatable {
        let achievedDays: Int
        let usedDays: Int

        var rate: Double {
            guard usedDays > 0 else { return 0 }
            return Double(achievedDays) / Double(usedDays)
        }
    }

    static func calculate(
        records: [WorkoutRecord],
        firstUseDate: Date,
        today: Date,
        calendar: Calendar = .mondayFirst
    ) -> Stats {
        let firstUseStart = calendar.startOfDay(for: firstUseDate)
        let todayStart = calendar.startOfDay(for: today)

        // Used days = inclusive day count from first use to today (min 1).
        let daysBetween = calendar.dateComponents([.day], from: firstUseStart, to: todayStart).day ?? 0
        let usedDays = max(1, daysBetween + 1)

        // Achieved days = unique days where at least one workout record was achieved.
        let achievedDates = records
            .filter { AchievementEvaluator.isAchieved(record: $0) }
            .map { calendar.startOfDay(for: $0.date) }
        let uniqueAchieved = Set(achievedDates).count

        return Stats(achievedDays: uniqueAchieved, usedDays: usedDays)
    }
}

File: app-android/app/src/main/java/com/goexercise/app/domain/LifetimeStatsCalculator.kt
package com.goexercise.app.domain

import java.time.LocalDate
import java.time.temporal.ChronoUnit

/**
 * 累計の達成日数・利用日数・達成率。iOS `LifetimeStatsCalculator.swift` の 1:1 移植。
 */
object LifetimeStatsCalculator {

    data class Stats(
        val achievedDays: Int,
        val usedDays: Int,
    ) {
        val rate: Double
            get() = if (usedDays > 0) achievedDays.toDouble() / usedDays.toDouble() else 0.0
    }

    fun calculate(
        records: List<WorkoutRecord>,
        firstUseDate: LocalDate,
        today: LocalDate,
    ): Stats {
        // 利用日数 = 初回利用日〜今日の inclusive(最低 1)。
        val daysBetween = ChronoUnit.DAYS.between(firstUseDate, today).toInt()
        val usedDays = maxOf(1, daysBetween + 1)

        // 達成日数 = 達成した記録のあるユニーク日数。
        val uniqueAchieved = records
            .filter { AchievementEvaluator.isAchieved(it) }
            .map { it.date }
            .toSet()
            .size

        return Stats(achievedDays = uniqueAchieved, usedDays = usedDays)
    }
}
